File: trees-and-binary-search/208-implement-trie-prefix-tree/implement-trie-prefix-tree.cpp
class Trie {
public:

    struct TrieNode {

        TrieNode* child[26];
        bool end;

        TrieNode() {

            end = false;

            for (int i = 0; i < 26; i++) {
                child[i] = nullptr;
            }
        }
    };

    TrieNode* root;

    Trie() {
        root = new TrieNode();
    }

    void insert(string word) {

        TrieNode* curr = root;

        for (char ch : word) {

            int idx = ch - 'a';

            if (curr->child[idx] == nullptr) {
                curr->child[idx] = new TrieNode();
            }

            curr = curr->child[idx];
        }

        curr->end = true;
    }

    bool search(string word) {

        TrieNode* curr = root;

        for (char ch : word) {

            int idx = ch - 'a';

            if (curr->child[idx] == nullptr) {
                return false;
            }

            curr = curr->child[idx];
        }

        return curr->end;
    }

    bool startsWith(string prefix) {

        TrieNode* curr = root;

        for (char ch : prefix) {

            int idx = ch - 'a';

            if (curr->child[idx] == nullptr) {
                return false;
            }

            curr = curr->child[idx];
        }

        return true;
    }
};
/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: trees-and-binary-search/208-implement-trie-prefix-tree/implement-trie-prefix-tree.java
class Trie {

    class TrieNode {
        TrieNode[] child = new TrieNode[26];
        boolean end;
    }

    TrieNode root;

    public Trie() {
        root = new TrieNode();
    }

    public void insert(String word) {

        TrieNode curr = root;

        for (char ch : word.toCharArray()) {

            int idx = ch - 'a';

            if (curr.child[idx] == null) {
                curr.child[idx] = new TrieNode();
            }

            curr = curr.child[idx];
        }

        curr.end = true;
    }

    public boolean search(String word) {

        TrieNode curr = root;

        for (char ch : word.toCharArray()) {

            int idx = ch - 'a';

            if (curr.child[idx] == null) {
                return false;
            }

            curr = curr.child[idx];
        }

        return curr.end;
    }

    public boolean startsWith(String prefix) {

        TrieNode curr = root;

        for (char ch : prefix.toCharArray()) {

            int idx = ch - 'a';

            if (curr.child[idx] == null) {
                return false;
            }

            curr = curr.child[idx];
        }

        return true;
    }
}